src/Qir/Runtime/lib/Simulators/FullstateSimulator.cpp

```diff
@@ -408,7 +419,109 @@ namespace Quantum
 
             return (std::abs(actualProbability - probabilityOfZero) < precision);
         }
-    };
+
+      private:
+        std::ostream& GetOutStream(const void* location, std::ofstream& outFileStream);
+        void DumpMachineImpl(std::ostream& outStream);
+        void DumpRegisterImpl(std::ostream& outStream, const QirArray* qubits);
+        void GetStateTo(TDumpLocation location, TDumpToLocationCallback callback);
+        bool GetRegisterTo(TDumpLocation location, TDumpToLocationCallback callback, const QirArray* qubits);
+
+      private:
+        TDumpToLocationCallback const dumpToLocationCallback = [](size_t idx, double re, double im, TDumpLocation location) -> bool
+            {
+                std::ostream& outStream = *reinterpret_cast<std::ostream*>(location);
+
+                if (re != 0 || im != 0)
+                {
+                    outStream << "|" << std::bitset<8>(idx) << ">: " << re << "+" << im << "i" << std::endl;
+                }
+                return true;
+            };
+    }; // class CFullstateSimulator
+
+    void CFullstateSimulator::DumpMachineImpl(std::ostream& outStream)
+    {
+        outStream << "*********************" << std::endl;
+        this->GetStateTo((TDumpLocation)&outStream, dumpToLocationCallback);
+        outStream << "*********************" << std::endl;
+        outStream.flush();
+    }
+
+    void CFullstateSimulator::DumpRegisterImpl(std::ostream& outStream, const QirArray* qubits)
+    {
+        outStream << "*********************" << std::endl;
+        if(!this->GetRegisterTo((TDumpLocation)&outStream, dumpToLocationCallback, qubits))
+        {
+            outStream << "## Qubits were entangled with an external qubit. Cannot dump corresponding wave function. ##" << std::endl;
+        }
+        outStream << "*********************" << std::endl;
+        outStream.flush();
+    }
+
+    bool CFullstateSimulator::GetRegisterTo(TDumpLocation location, TDumpToLocationCallback callback, const QirArray* qubits)
+    {
+        // `qubits->buffer` points to the sequence of qubit ids each of type `void*` (logically `qubits->buffer` is of type `void**`).
+        // We need to pass to the native simulator the pointer to the sequence of `unsigned`.
+        std::vector<unsigned> unsQbIds(qubits->count);
+        size_t i = 0;
+        for (unsigned& qbId : unsQbIds)
+        {
+            qbId = (unsigned)(uintptr_t)(((void**)(qubits->buffer))[i]);

```

Comment:
It feels a little odd to see as many C-style casts here all of a sudden, is there a reason for that?

Reply:
Yep, otherwise the compiler complains. :-D

1. (regarding the C-style casts as opposed to C++-style casts)
The C-style casts are shorter and more readable than the C++-style casts (when we are comparing with `reinterpret_cast<>()`).

2. (regarding the number of casts) 

By C++ definition, the type of `qubits->buffer` is `char *` (pointer to the sequence of `char`s).

In reality the `qubits->buffer` (for an array of _qubits_) points to the sequence of qubit IDs, each qubit ID having a size of a pointer (i.e. the size of qubit ID is currently implementation-dependent, and each qubit ID gets cast to, and stored as, a `void*`). To summarize, the `qubits->buffer` points to the sequence of pointers to `void` i.e. the real type of the `qubits->buffer` is `void**` (hence the cast to `void**`).
Then we index that pointer - `[i]` - i.e. we get the corresponding `void*` out of the sequence. On a 64-bit platform that `void*` is 64-bit. Eventually we need to cast it to `unsigned` (which in current (compiler) implementation is 32-bit). A direct _narrowing_ cast from a 64-bit _pointer_ to a 32-bit _integer_ causes a Clang compiler error or warning (regardless of whether the cast is implicit or explicit, and whether it is a C-style cast or C++-style cast). That is why I have to use an intermediate cast to an integer type -  [`uintptr_t`](https://man7.org/linux/man-pages/man0/stdint.h.0p.html) - whose size is the same as the pointer type (regardless of whether the architecture is 64-bit or 32-bit). And then I apply the _narrowing_ cast from 64-bit _integer_ to 32-bit _integer_ (`unsigned`), and the compiler is happy.

May be the last cast to `unsigned` can be _implicit_, I didn't check or don't remember, but the static analyzers typically complain for the _implicit narrowing_ casts. So I have cast explicitly.

Reply:
I wonder, instead of copying the qubit id vector, couldn't we just pass the buffer directly to the simulator? something like:
`return dumpQubitsToLocation(this->simulatorId, (unsigned)(qubits->count), (unsigned*)(qubits->buffer), callback, location);`
Is there really anything to be gained by converting the ids one at a time instead of all at once?

Reply:
@swernli, thanks for admitting.
What you are suggesting, was my original attempt. But `dumpQubitsToLocation()` expects as the 3rd argument a pointer to the sequence of `unsigned` (32-bit each), but the `qubits->buffer` is a pointer to the sequence of `void*` (64-bt each). We have to convert.
Some more info on the topic - #631

Reply:
ah, I understand now. Yeah, that is annoying, but this appears to be the best way to accomplish it for now. Thanks for explaining!